### PR TITLE
Openid connect -> main

### PR DESCRIPTION
**Pull request by npalaska**
Since we have now cut a deployment branch that's `0.72`, its time to bring the changes from the `openid-connect` branch on the main and sync with the main.

### CODE REVIEW

File: lib/pbench/server/api/resources/users_api.py
Comment (dbutenhof) [inline lib/pbench/server/api/resources/users_api.py:1]: Why does this show as a *diff* instead of a removal? Do we have an empty `users_api.py`??

File: lib/pbench/server/database/models/active_tokens.py
Comment (dbutenhof) [inline lib/pbench/server/database/models/active_tokens.py:1]: Adding an empty file here ... we need to resolve `auth_tokens` vs `active_tokens`. "There can be only one!" (I.e., is anyone importing this?)

File: lib/pbench/test/functional/server/test_put.py
Comment (dbutenhof) [inline lib/pbench/test/functional/server/test_put.py:337]: This hack bothers me. Fix the fixture to restore the previous function scope login behavior, which was intended to support test cases like this!

File: lib/pbench/test/functional/server/conftest.py
Comment (dbutenhof) [inline lib/pbench/test/functional/server/conftest.py:64]: Ah: that's what's happening. `server_client` is a *session* scope, and you removed the `yield` and `logout` from the fixture.

Restore the `yield` and just clear `server_client.auth_token` afterwards, and you'll restore the previous "per test case" login, which allows the non-login test cases to run without hacks.

**Review by dbutenhof — COMMENTED** [review]
I wonder if you should hold off here until you've got the dashboard `Keycloak` package and hooks integrated? This will break the dashboard until that's done. But maybe that's OK ... ❔ 👀 

In any case, there seem to be a few empty files, and a few other possible problems ...

**Review by webbnh — COMMENTED** [review]
Here's what I've got so far...have a good weekend!

File: dashboard/src/actions/authActions.js
Comment (webbnh) [inline dashboard/src/actions/authActions.js:36]: What is supposed to happen if `error?.response` is false-y?  (I don't see a reducer which handles type `OPENID_ERROR`...did I miss it?)

**Review by webbnh — CHANGES_REQUESTED** [review]
OK, the first pass is done.  Bring on the updates!

File: lib/pbench/client/__init__.py
Comment (npalaska) [reply lib/pbench/client/__init__.py:341]: All the error handling is done by the `user_login` method of the `OIDCAdmin` class inside the `oidc_admin.py` file. The class inherits from `pbench.server.auth.Connection` andthis Connection class has all the required error handling baked into it.

File: lib/pbench/client/__init__.py
Comment (npalaska) [reply lib/pbench/client/__init__.py:345]: No requirement for it yet.

File: lib/pbench/server/database/models/auth_tokens.py
Comment (npalaska) [reply lib/pbench/server/database/models/auth_tokens.py:26]: Siddardh is adding it as a part of the work on PBENCH-1067 but I am just treating it as if there is no tokens table yet.

File: lib/pbench/client/__init__.py
Comment (npalaska) [reply lib/pbench/client/__init__.py:443]: Even better, I am gonna remove it as its not in use anymore.

File: lib/pbench/client/__init__.py
Comment (npalaska) [reply lib/pbench/client/__init__.py:447]: Removing this function.

File: lib/pbench/client/oidc_admin.py
Comment (npalaska) [reply lib/pbench/client/oidc_admin.py:39]: `OIDCAdmin` class inherits from the `Connection` class and it has all the required error handling there, here I am just propagating the exception raised by the post.

File: server/pbenchinacan/load_keycloak.sh
Comment (npalaska) [reply server/pbenchinacan/load_keycloak.sh:27]: > The Pbench Dashboard will not (we assume...) be the only client of the Pbench Server...so I don't think naming it here is appropriate. If we don't like `"pbench-server-client"`, then how about `"pbench-client"`?

I didn't understand, What do you mean?

File: server/lib/config/pbench-server-default.cfg
Comment (npalaska) [reply server/lib/config/pbench-server-default.cfg:115]: Ohh do you mean client as in the _OIDC_ client or the client as in the _user_ the Pbench-Server?

File: lib/pbench/server/database/alembic/versions/f628657bed56_user_table_update_oidc.py
Comment (npalaska) [reply lib/pbench/server/database/alembic/versions/f628657bed56_user_table_update_oidc.py:31]: > Why do we pass existing_nullable in the first call

`existing_nullable` operation indicates whether a column that already exists in the database can be set to allow null values.

The `existing_nullable` argument comes into play when you want to modify an existing column that already has data in it. If we set `existing_nullable=True`, this means that the column will be set to allow null values if it doesn't already, and any existing data in the column will not be affected. If we set `existing_nullable=False`, this means that the column will be set to disallow null values if it currently allows them and any existing data that is null will cause an error. This is because the column is no longer allowed to have null values, so the existing null values are now considered invalid data. _I think_ that's what we want for the `id` column (as its a primary key) and we never expect it to have a `null` value if it does then it should result in an error. (Now, when we deploy this we would create a new users table so the scenario wouldn't matter but I think this is a good check to have.) 

> but not the second

`existing_nullable` does not exist in the `username` column because I think for the `username` column Alembic does not need to change the existing nullability of the column. In other words, if the column currently allows null values, it will continue to allow null values after the migration, and if it currently disallows null values, it will continue to disallow null values.

> why do we pass nullable in the second but not the first

We aren't specifying `nullable=False` because we added a primary key constraint on `id` column so its not needed.

File: lib/pbench/server/database/models/users.py
Comment (npalaska) [reply lib/pbench/server/database/models/users.py:166]: Yeah, we don't have a consumer for this `delete` yet but I think we do want to provide this functionality eventually if the user wishes to delete himself/herself from the Pbench account. 

> I don't see any code here which prevents an admin user from being deleted.

I should update the comment. The concept of an admin user on the Pbench database is no longer there so this comment is no longer relevant.

File: lib/pbench/test/unit/server/auth/test_auth.py
Comment (npalaska) [reply lib/pbench/test/unit/server/auth/test_auth.py:584]: Yeah, need to update the docstring

File: lib/pbench/test/unit/server/conftest.py
Comment (npalaska) [reply lib/pbench/test/unit/server/conftest.py:145]: Yeah, you are right but I guess the test ensures that the code can correctly encrypt and decrypt messages using an RSA key-pair. While we could use a mock object to simulate the behavior of the key-pair, this would not guarantee that the code can work correctly with a real RSA key-pair. I am just moving this code that Peter added earlier, if we want to update this to a mock object I can open an issue work on it post this PR.

File: lib/pbench/test/unit/server/database/test_users_db.py
Comment (npalaska) [reply lib/pbench/test/unit/server/database/test_users_db.py:113]: Yeah that's one of the things that need some expanding on `fake_db`, it does not handle table relationships yet.

File: lib/pbench/test/unit/server/test_user_auth.py
Comment (npalaska) [reply lib/pbench/test/unit/server/test_user_auth.py:1]: Yeah it got added as empty file as part of rebase, will remove it.

File: lib/pbench/test/unit/server/database/__init__.py
Comment (npalaska) [reply lib/pbench/test/unit/server/database/__init__.py:401]: Yeah, I recall being hesitant to make this change due to the exact concerns you pointed out about function performance. However, I subsequently made the change to address some errors that were being raised testing with `fake_db`. Subsequently, I made several additional changes to the code and overlooked the need to retest this specific modification.

Upon retesting the code using the `sorted` method, all tests seem to pass now. I will revert this change.

File: server/lib/config/pbench-server-default.cfg
Comment (webbnh) [reply server/lib/config/pbench-server-default.cfg:104]: See [above](https://github.com/distributed-system-analysis/pbench/pull/3353#discussion_r1149602916).

File: lib/pbench/server/database/alembic/versions/f628657bed56_user_table_update_oidc.py
Comment (webbnh) [reply lib/pbench/server/database/alembic/versions/f628657bed56_user_table_update_oidc.py:31]: Thanks for the explanation!

So, in the first case we don't need to specify `nullable=False` (and so we don't) because that's implied by the fact that the `id` column already is the primary key.  And, on that basis, the `existing_nullable=False` specification should not find any issues, but perhaps it's prudent to have it look.

The second case is odd, because there's no change in the source code to this column...so why do we need a migration alteration for it?  (It's already a non-nullable `String(255)`.)

File: lib/pbench/server/database/models/users.py
Comment (webbnh) [reply lib/pbench/server/database/models/users.py]: > I think we do want to provide this functionality eventually if the user wishes to delete himself/herself from the Pbench account.

This database table is an implementation detail of the Pbench Server.  It contains only cached data gleaned from the Keycloak token, and the data will be updated each time the user logs into the Pbench Server.  There's really nothing here for the user to need to delete.  (Conceivably, we could purge it on a periodic basis or when the user logs out.)

However, we don't want the username-to-ID correlation removed as long as there are existing datasets for the user.  So, before we allow this delete operation to go forward, we need to ensure that all the user's datasets have either been removed or transferred to other ownership.

This deletion operation can easily be performed by a human using direct SQL commands on the database.  There is no need to provide a programmatic method for doing this, and the code here is incomplete.

I really think you should omit this function from the PR.

File: server/pbenchinacan/load_keycloak.sh
Comment (npalaska) [reply server/pbenchinacan/load_keycloak.sh:27]: So instead of `pbench-dashboard` we should name our client as more generic like `pbench-client`??

File: lib/pbench/test/unit/server/auth/test_auth.py
Comment (webbnh) [inline lib/pbench/test/unit/server/auth/test_auth.py:594]: Should this be `"test-verify-auth-oidc-user-update"`?

**Review by webbnh — COMMENTED** [review]
Coming along nicely.

The new parameter for the `_decode()` function looks like it shouldn't be optional/`None`, and it should be specified in the new calls.

The name of the ODIC client in the `Pbench` realm should be something more generic than `pbench-dashboard`...unless we're planning on configuring custom clients for all of our programmatic users....

Other than those, my new comments are just nits and small stuff.

There are a couple of comments from my previous reviews which you missed or need attention:

- [x] [Docstring update](https://github.com/distributed-system-analysis/pbench/pull/3353#discussion_r1142780784).
- [x] [Dashboard error handling](https://github.com/distributed-system-analysis/pbench/pull/3353#discussion_r1143984271).
- [x] The `User` table [delete](https://github.com/distributed-system-analysis/pbench/pull/3353#discussion_r1149659130) function.
- [ ] The [`"username"` column](https://github.com/distributed-system-analysis/pbench/pull/3353#discussion_r1149646630) in the Alembic migration code.

File: lib/pbench/server/database/models/users.py
Comment (npalaska) [reply lib/pbench/server/database/models/users.py]: Ah! that's correct.

File: lib/pbench/server/database/alembic/versions/f628657bed56_user_table_update_oidc.py
Comment (npalaska) [reply lib/pbench/server/database/alembic/versions/f628657bed56_user_table_update_oidc.py:31]: That's a good point. I have no idea why `alembic` decided to add this line in the migration file, I'll have to do some digging.

File: dashboard/src/actions/authActions.js
Comment (npalaska) [reply dashboard/src/actions/authActions.js]: With the recent rebase this function should be gone.

File: lib/pbench/test/functional/server/test_put.py
Comment (dbutenhof) [reply lib/pbench/test/functional/server/test_put.py:337]: You fixed the fixture, so this should go ...

**Review by dbutenhof — COMMENTED** [review]
Since this appears to be still in draft, I just made a scan to follow up on my previous comments, most of which (save one) seem to be resolved...

**Review by dbutenhof — COMMENTED** [review]
Other than the dangling endpoints client reference that blew the tests, this is looking good.

**Review by dbutenhof — APPROVED** [review]

File: lib/pbench/server/database/alembic/versions/f628657bed56_user_table_update_oidc.py
Comment (npalaska) [reply lib/pbench/server/database/alembic/versions/f628657bed56_user_table_update_oidc.py:31]: I did some testing around this and alembic doesn't seem to complain when I remove this alteration, nor it has any problem building tables. So I guess can remove it.

**Review by dbutenhof — APPROVED** [review]
I guess if Mr Jenkins is happy, I'm happy.

File: dashboard/src/actions/overviewActions.js
Comment (webbnh) [inline dashboard/src/actions/overviewActions.js:52]: No need to change this unless you have other changes to make, but a recent change from Varshini brought to light the fact that we have [alternatives](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing#description) to the use of the ternary operator, here.  For instance, either `||` or `??` might be nicer.
```javascript
      dispatch(showToast(DANGER, msg || `Error response: ERROR_MSG`));
```
(`??` will use the default message if `msg` is `undefined`; `||` will use the default message if `msg` is an empty string; but using either of them allows us to avoid the second reference to `msg`...just some more items for our Javascript toolboxes....)

File: lib/pbench/server/database/alembic/versions/f628657bed56_user_table_update_oidc.py
Comment (webbnh) [reply lib/pbench/server/database/alembic/versions/f628657bed56_user_table_update_oidc.py]: Have you tried regenerating this code?  (I'm guessing that the regeneration would omit this alteration, now.)

I assume that the Alembic "check" test would catch any divergence between a migration script generated early in a PR's evolution and subsequent incompatible changes in the PR, but...as a best practice, we should probably generate the migration script late in the PR process (or, perhaps, _re_-generate it periodically during review...if there are any relevant changes requested).

File: dashboard/src/actions/overviewActions.js
Comment (npalaska) [reply dashboard/src/actions/overviewActions.js:52]: Okay, I don't have any blocking changes outstanding in this PR now so I'll defer it in another PR.

**Review by webbnh — APPROVED** [review]
This looks good to go.

I found two more nits after the earlier one that GitHub helpfully posted before I I was ready, and there's one more which I think I called out before but cannot find the comment for...but none of these need to gate the merge of this PR.

File: dashboard/src/actions/authActions.js
Comment (npalaska) [reply dashboard/src/actions/authActions.js:71]: Yeah in this PR I just sort of followed how the functions were written already in this file without giving a lot of thought to javascript async operations (which I now have a better understanding of than before I started working on #3233 and #3250). But if this is not a blocking change then I'll defer to my next dashboard PR (Removal of the user profile page).

File: lib/pbench/test/unit/server/database/__init__.py
Comment (npalaska) [reply lib/pbench/test/unit/server/database/__init__.py:316]: Sigh! I'll fix it in the next PR.

**Review by webbnh — APPROVED** [review]
No visible changes 🙈, it must be good.  👍

**Pull request merged**
